pinot-minion/src/main/java/org/apache/pinot/minion/taskfactory/TaskFactoryRegistry.java

```diff
@@ -88,24 +104,24 @@ public TaskResult run() {
                 Object executionResult = _taskExecutor.executeTask(pinotTaskConfig);
                 long timeSpentInNanos = System.nanoTime() - startTimeInNanos;
                 _eventObserver.notifyTaskSuccess(pinotTaskConfig, executionResult);
-                minionMetrics.addMeteredTableValue(taskType, MinionMeter.NUMBER_TASKS_COMPLETED, 1L);
-                minionMetrics.addPhaseTiming(taskType, MinionQueryPhase.TASK_EXECUTION, timeSpentInNanos);
+                _minionMetrics.addMeteredTableValue(taskType, MinionMeter.NUMBER_TASKS_COMPLETED, 1L);
+                _minionMetrics.addPhaseTiming(taskType, MinionQueryPhase.TASK_EXECUTION, timeSpentInNanos);
```

Comment:
move this to the run() method

Reply:
@mcvsubbu I am a little bit confused. Currently `runInternal()` just simply call `Object executionResult = _taskExecutor.executeTask(pinotTaskConfig);` then try/catch all kinds of exceptions, and log/emit metric accordingly. If we add try/catch in `run()` and move the log/metric logic to the `run()` method, then it's equivalent to moving all logic to`run()`:

```java
   public TaskResult run() {
     try {
       Object executionResult = _taskExecutor.executeTask(pinotTaskConfig);
       // log and emit success message/metric
       return new TaskResult(TaskResult.Status.COMPLETED, "Succeeded");
     } catch (TaskCancelledException e) {
       // log and emit cancel message/metric
       return new TaskResult(TaskResult.Status.CANCELED, e.toString());
     } catch (FatalException e) {
       // log and emit fatal error message/metric
       return new TaskResult(TaskResult.Status.FATAL_FAILED, e.toString());
     } catch (Exception e) {
        // log and emit error message/metric
       return new TaskResult(TaskResult.Status.FAILED, e.toString());
     } finally {
       _minionMetrics.addValueToGlobalGauge(MinionGauge.NUMBER_OF_TASKS, -1L);
     }
   }
   private Object runInternal(PinotTaskConfig pinotTaskConfig)
       throws Exception {
       return  _taskExecutor.executeTask(pinotTaskConfig)
   }

```

Reply:
For now, runInternal() seems to catch all exceptions, yes. And we can leave it like that.
In the parent run() method, I suggest:
try {
   runInternal()
} finally {
   update metrics
}

This will protect against any changes inside the runInternal logic that may choose to bubble up exceptions or if there is some uncaught exception by mistake.

Reply:
done